File: src/contests/cg/puzzle/java/fibonaccis-rabbit.java
// 🎮 CodinGame Puzzle - fibonaccis-rabbit
// https://www.codingame.com/training/easy/fibonaccis-rabbit

import java.io.*;
import java.math.BigInteger;

class Solution {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] l1 = br.readLine().trim().split(" ");
        int f0 = Integer.parseInt(l1[0]), n = Integer.parseInt(l1[1]);
        String[] l2 = br.readLine().trim().split(" ");
        int a = Integer.parseInt(l2[0]), b = Integer.parseInt(l2[1]);
        // FN can exceed 2^63 (but is < 2^64), so accumulate with BigInteger.
        BigInteger[] f = new BigInteger[n + 1];
        for (int i = 0; i <= n; i++) f[i] = BigInteger.ZERO;
        f[0] = BigInteger.valueOf(f0);
        for (int i = 1; i <= n; i++) {
            BigInteger total = BigInteger.ZERO;
            for (int k = a; k <= b; k++) {
                if (i - k >= 0) total = total.add(f[i - k]);
            }
            f[i] = total;
        }
        System.out.println(f[n]);
    }
}

File: src/contests/cg/puzzle/fsharp/fibonaccis-rabbit.fs
// 🎮 CodinGame Puzzle - fibonaccis-rabbit
// https://www.codingame.com/training/easy/fibonaccis-rabbit

let l1 = System.Console.ReadLine().Trim().Split(' ')
let f0 = int l1.[0]
let n = int l1.[1]
let l2 = System.Console.ReadLine().Trim().Split(' ')
let a = int l2.[0]
let b = int l2.[1]
// FN can exceed 2^63 (but is < 2^64), so accumulate in uint64.
let f = Array.zeroCreate<uint64> (n + 1)
f.[0] <- uint64 f0
for i in 1 .. n do
    let mutable total = 0UL
    for k in a .. b do
        if i - k >= 0 then total <- total + f.[i - k]
    f.[i] <- total
printfn "%d" f.[n]
